doc/aaa/RBAC_Phase2.md

```diff
@@ -0,0 +1,421 @@
+
+## 3.5 SAI
+N/A
+
+## 3.6 User Interface
+### 3.6.1 Data Models
+TBD from developer
+(TODO/DELL)
+
+### 3.6.2 CLI
+#### 3.6.2.1 Configuration Commands for User Management
+Users may be managed via Linux tools like `useradd`, `usermod`, `passwd`, etc. They may also be managed via configuration.
+
+##### username
+`username <name> password <password-string> role <role-string>` -- Configures a user on the system with a given name, password, and role.
+* **name** is a text string of 1-32 alphanumeric characters
+* **password-string** is a text string of 1-32 alphanumeric characters
+* **role-string** is a text string consisting of a role name. In the initial release, the user is recommended to use "admin" and "operator" roles, as other roles will not be supported. A text string is desired instead of keywords so that in the future, more roles may be implemented and expanded.
+* Configuring another a user with the same **name** should result in modification of the existing user.
+
+`no username <name>` -- Deletes a user from the system.
+* **name** is a text string of 1-32 alphanumeric characters
+
+#### 3.6.2.2 Show Commands
+N/A
+
+#### 3.6.2.3 Debug Commands
+N/A
+
+#### 3.6.2.4 IS-CLI Compliance
+N/A
+
+### 3.6.3 REST API Support
+Ability to configure local users via REST API.
+
+# 4 Flow Diagrams
+N/A
+
+# 5 Error Handling
+## 5.1 REST Server
+The REST server should return standard HTTP errors when authentication fails or if the user tries to access a forbidden resource or perform an unauthorized activity.
+
+## 5.2 gNMI server
+The gNMI server should return standard gRPC errors when authentication fails.
+
+## 5.3 CLI
+Authentication errors will be handled by SSH. However, the CLI must gracefully handle authorization failures from the REST server (the authorization failure would originate from Translib of course). While the CLI will render all of the available commands to a user, the user will actually only be able to execute a subset of them. This limitation is a result of the design decision to centralize RBAC in Translib. Nevertheless, the CLI must inform the user when they attempt to execute an unauthorized command.
+
+## 5.4 Translib
+Translib will authorize the user and when the authorization fails will return appropriate error string to the REST/gNMI server.
+
+If a user authenticates but is not part of one of the pre-defined groups, they will not be allowed to do anything at all on the system.
+
+# 6 Serviceability and Debug
+All operations performed by NBIs (CLI commands, REST/gNMI operations) should be logged/audited with usernames attached to the given operation(s) performed.
+
+Initially, users who are remotely authenticated will share a common role-specific username, so there will be a limitation here.
+
+# 7 Warm Boot Support
+N/A
+
+# 8 Scalability
+See previous section 1.1.3: Scalability Requirements
+
+# 9 Unit Test
+
+### Table 3: Test Cases
```

Comment:
Will need to update the list of test cases to include custom roles and users with multiple roles and stuff